### DISCUSSION TITLE
Multiple shairport-sync instances in LXC containers

### DISCUSSION THREAD

**Discussion by noelhibbard** [mikebrady/shairport-sync#1250] [Show and tell]
# Deploy shairport-sync in Debian LXC container
I loaded Debian Buster on bare metal to compose these instructions but they should work on other Debian based distros like Ubuntu but I haven't tested. These instructions also explain how to map the sound cards into the containers so it's not necessary to use a PulseAudio server. Commands prefixed with a # will are run on the host. Commands prefixed with a $ will be run in the LXC container.

## Install LXC and ALSA
```
# apt update
# apt install -y lxc alsa-utils
```

## Create network bridge
You need to create a network bridge in order for the containers to sit on your host network and aquire an IP via DHCP. Follow these directions: https://wiki.debian.org/LXC/SimpleBridge#Host_device_as_bridge

## Reboot Server after creating bridge
You can restart the network without rebooting but you risk loosing connectivity so it's safer to just reboot
```
# reboot
```

## Create new LXC container

```
# lxc-create --name <containername> --template debian
```

Configure the container so that it has access to your soundcards and uses the network bridge created above.
```
# nano /var/lib/lxc/<containername>/config
```

Add/change these lines:
```
lxc.start.auto = 1
lxc.net.0.type = veth
lxc.net.0.link = lxcbr0
lxc.cgroup.devices.allow = c 116:* rwm
lxc.mount.entry = /dev/snd dev/snd none bind,optional,create=dir
```

If you have an elaborate ALSA config you can map the ALSA config file into the container so you only have to maintain the config on the host. To do that, add a line like this to the LXC config file:
```
lxc.mount.entry = /etc/asound.conf etc/asound.conf none bind,optional,create=file
```

## Start container, set the root password and enable root access via SSH
```
# lxc-start <containername>
# lxc-attach <containername>
$ passwd
$ sed -i 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/g' /etc/ssh/sshd_config
$ systemctl restart sshd
$ exit
```

# Now you can SSH into the container
```
# ssh root@<containername>
```

# Install prereqs for compiling/running shairport-sync
```
$ apt update
$ apt upgrade
$ apt install -y xxd alsa-utils nano build-essential sudo git xmltoman autoconf automake libtool \
    libpopt-dev libconfig-dev libasound2-dev avahi-daemon libavahi-client-dev libssl-dev libsoxr-dev \
    libglib2.0-dev libplist-dev libsodium-dev libavutil-dev libavcodec-dev libavformat-dev uuid-dev \
    libgcrypt-dev libmosquitto-dev libsndfile1-dev

$ git clone https://github.com/mikebrady/nqptp.git ~/nqptp
$ cd ~/nqptp
$ autoreconf -fi
$ ./configure
$ make
$ make install
$ systemctl enable nqptp
$ systemctl start nqptp

$ git clone https://github.com/mikebrady/alac.git ~/alac
$ cd ~/alac
$ autoreconf -fi
$ ./configure
$ make
$ make install

$ git clone https://github.com/mikebrady/shairport-sync.git -b development ~/shairport-sync
$ cd ~/shairport-sync
$ autoreconf -fi
$ ./configure --sysconfdir=/etc --with-alsa \
    --with-soxr --with-ssl=openssl --with-avahi --with-airplay-2 \
    --with-mqtt-client --with-alsa --with-convolution --with-systemd --with-apple-alac
$ make -j
$ make install
```

## Edit shairport-sync.conf however you wish (the defaults should work fine though)
```
$ nano /etc/shairport-sync.conf
```

## Enable/Start shairport-sync service
```
$ systemctl enable shairport-sync
$ systemctl start shairport-sync
```

## Other tidbits
Once you have the container the way you like it, you can make a copy with another name to run another instance of shairport-sync. The new container will have a random MAC address so it will get its own IP.
```
# lxc-copy -n <source_container_name> -N <destination_container_name>
```
The root file system for a container it located here. You can copy files between the host/container and edit files from the host but it's a good idea to do this while the container is in a stopped state:
```
/var/lib/lxc/<containername>/rootfs
```

The containers should autostart when the host starts. If you need to stop all the containers you can use this command
```
# lxc-autostart --shutdown
```

This command will autostart any containers that aren't already running:
```
# lxc-autostart
```

**Comment by mattvv** [comment]
Awesome guide! thank you!

**Comment by DaCHack** [comment]
Hi, @noelhibbard ,

many thanks for this write-up. I am trying to use this for a Proxmox LXC on my Raspberry Pi 4 and experience issues.
Any idea what I am doing wrong?

- Using Proxmox 7 (Pimox7) on Raspberry Pi 4 2GB ("HyperVi")
- Setup a privileged Proxmox LXC ("MediaCenter") via the web interface using a Debian template
- Edited the container config files as described in your guide
- Both host and LXC container do not have a asound.conf so I left this step out
- On the host, aplay sees my 2 HDMI connectors as audio devices and speaker-test gives an output through the speakers
- On the LXC, aplay does not see any audio device

```
root@HyperVi:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: vc4hdmi0 [vc4-hdmi-0], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: vc4hdmi1 [vc4-hdmi-1], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

```
root@MediaCenter:~# aplay -l
aplay: device_list:274: no soundcards found...
```

```
root@HyperVi:~# ls -lash /dev/snd
total 0
0 drwxr-xr-x  3 root root      180 Dec  1 22:53 .
0 drwxr-xr-x 18 root root     4.1K Dec  1 22:54 ..
0 drwxr-xr-x  2 root root       80 Dec  1 22:54 by-path
0 crw-rw----  1 root audio 116,  0 Dec  1 22:53 controlC0
0 crw-rw----  1 root audio 116, 32 Dec  1 22:54 controlC1
0 crw-rw----  1 root audio 116, 16 Dec  1 22:53 pcmC0D0p
0 crw-rw----  1 root audio 116, 48 Dec  1 22:54 pcmC1D0p
0 crw-rw----  1 root audio 116,  1 Dec  1 22:53 seq
0 crw-rw----  1 root audio 116, 33 Dec  1 22:53 timer
```

```
root@MediaCenter:~# ls -lash /dev/snd/
total 0
0 drwxr-xr-x 3 root root      180 Dec  1 21:53 .
0 drwxr-xr-x 7 root root      480 Dec  1 21:54 ..
0 drwxr-xr-x 2 root root       80 Dec  1 21:54 by-path
0 crw-rw---- 1 root audio 116,  0 Dec  1 21:53 controlC0
0 crw-rw---- 1 root audio 116, 32 Dec  1 21:54 controlC1
0 crw-rw---- 1 root audio 116, 16 Dec  1 21:53 pcmC0D0p
0 crw-rw---- 1 root audio 116, 48 Dec  1 21:54 pcmC1D0p
0 crw-rw---- 1 root audio 116,  1 Dec  1 21:53 seq
0 crw-rw---- 1 root audio 116, 33 Dec  1 21:53 timer
```

HDMI "soundcards" are showing in the container under proc:
```
root@MediaCenter:~# cat /proc/asound/cards
 0 [vc4hdmi0       ]: vc4-hdmi - vc4-hdmi-0
                      vc4-hdmi-0
 1 [vc4hdmi1       ]: vc4-hdmi - vc4-hdmi-1
                      vc4-hdmi-1
```

**/var/lib/lxc/100/config on the host:**
```
lxc.cgroup.relative = 0
lxc.cgroup.dir.monitor = lxc.monitor/100
lxc.cgroup.dir.container = lxc/100
lxc.cgroup.dir.container.inner = ns
lxc.arch = arm64
lxc.include = /usr/share/lxc/config/debian.common.conf
lxc.apparmor.profile = generated
lxc.apparmor.raw = deny mount -> /proc/,
lxc.apparmor.raw = deny mount -> /sys/,
lxc.monitor.unshare = 1
lxc.tty.max = 2
lxc.environment = TERM=linux
lxc.uts.name = MediaCenter
lxc.cgroup2.memory.max = 536870912
lxc.cgroup2.memory.swap.max = 536870912
lxc.rootfs.path = /var/lib/lxc/100/rootfs
lxc.net.0.type = veth
lxc.net.0.veth.pair = veth100i0
lxc.net.0.hwaddr = EA:4D:1D:96:D8:29
lxc.net.0.name = eth0
lxc.net.0.script.up = /usr/share/lxc/lxcnetaddbr
lxc.cgroup.devices.allow = c 116:* rwm
lxc.mount.entry = /dev/snd dev/snd none bind,create=dir 0 0
lxc.mount.entry = /dev/dri dev/dri none bind,create=dir 0 0
lxc.mount.entry = /dev/snd dev/snd none bind,create=dir 0 0
lxc.cgroup2.cpuset.cpus =
```

**Comment by reitermarkus** [comment]
@DaCHack, try changing `lxc.cgroup.devices.allow` to `lxc.cgroup2.devices.allow`. I had the same problem and I think that fixed it.

**Reply by DaCHack** [reply]
Thanks! I changed the setup to a docker-only style, but will change back to LXC eventually and try this approach. Makes sense once you know about the change from cgroup to cgroup2 :)

**Comment by lollilol** [comment]
for the nqptp part i would add 
`--with-systemd-startup` to the `./configure` command
> $ ./configure --with-systemd-startup
